sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/NonStreamingOrderByUtils.java

```diff
@@ -0,0 +1,105 @@
+        @SuppressWarnings("unchecked")
+        Flux<OrderByRowResult<Document>>[] fluxes = documentProducers
+            .subList(0, documentProducers.size())
+            .stream()
+            .map(producer ->
+                toNonStreamingOrderByQueryResultObservable(producer, tracker, queryMetricsMap, initialPageSize,
+                    consumeComparer, clientSideRequestStatistics))
+            .toArray(Flux[]::new);
+        return Flux.mergeOrdered(consumeComparer, fluxes);
+    }
+
+    private static Flux<OrderByRowResult<Document>> toNonStreamingOrderByQueryResultObservable(DocumentProducer<Document> producer,
+                                                                                               RequestChargeTracker tracker,
+                                                                                               Map<String, QueryMetrics> queryMetricsMap,
+                                                                                               int initialPageSize,
+                                                                                               OrderbyRowComparer<Document> consumeComparer,
+                                                                                               Collection<ClientSideRequestStatistics> clientSideRequestStatisticsList) {
+        return producer
+            .produceAsync()
+            .transformDeferred(new NonStreamingOrderByUtils.PageToItemTransformer(tracker, queryMetricsMap, initialPageSize,
+                consumeComparer, clientSideRequestStatisticsList));
+    }
+
+    private static class PageToItemTransformer implements
+        Function<Flux<DocumentProducer<Document>.DocumentProducerFeedResponse>, Flux<OrderByRowResult<Document>>> {
+        private final RequestChargeTracker tracker;
+        private final Map<String, QueryMetrics> queryMetricsMap;
+        private final Integer initialPageSize;
+        private final OrderbyRowComparer<Document> consumeComparer;
+        private final Collection<ClientSideRequestStatistics> clientSideRequestStatistics;
+
+        private PageToItemTransformer(RequestChargeTracker tracker, Map<String, QueryMetrics> queryMetricsMap,
+                                      Integer initialPageSize, OrderbyRowComparer<Document> consumeComparer,
+                                      Collection<ClientSideRequestStatistics> clientSideRequestStatistics) {
+            this.tracker = tracker;
+            this.queryMetricsMap = queryMetricsMap;
+            this.initialPageSize = initialPageSize;
+            this.consumeComparer = consumeComparer;
+            this.clientSideRequestStatistics = clientSideRequestStatistics;
+        }
+
+        @Override
+        public Flux<OrderByRowResult<Document>> apply(Flux<DocumentProducer<Document>.DocumentProducerFeedResponse> source) {
+            PriorityQueue<OrderByRowResult<Document>> priorityQueue = new PriorityQueue<>(consumeComparer);
```

Comment:
This is not thread-safe - flatMap can execute on multiple threads concurrently. It should actually also limit concurrency I guess?

Reply:
As discussed offline, using `PriorityBlockingQueue` for this. And using the flatMap with 1 as shown below, should take care of concurrency.

`return source.flatMap(documentProducerFeedResponse -> { "PQ logic here" },1)`